src/System.Numerics.Tensors/System/Numerics/Tensor.cs

```diff
@@ -0,0 +1,1197 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Diagnostics;
+
+namespace System.Numerics
+{
+
+    public static partial class Tensor
+    {
+        internal static void ValidateBinaryArgs<T>(Tensor<T> left, Tensor<T> right)
+        {
+            if (left == null)
+            {
+                throw new ArgumentNullException(nameof(left));
+            }
+
+            if (right == null)
+            {
+                throw new ArgumentNullException(nameof(right));
+            }
+
+            // TODO: Support "compatible sizes" AKA "broadcasting"
+
+            if (left.Rank != right.Rank || left.Length != right.Length)
+            {
+                throw new ArgumentException("Operands must have matching dimensions");
+            }
+
+            for (int i = 0; i < left.Rank; i++)
+            {
+                if (left.Dimensions[i] != right.Dimensions[i])
+                {
+                    throw new ArgumentException("Operands must have matching dimensions");
+                }
+            }
```

Comment:
This is O(numDimensions), whereas the operations are O(numElements) at best.  I don't expect this to make much difference but if it does we can try to come up with some better fast paths which "pack" the dimensions into a single value (rather than array) and allow for a fast comparison for smaller numbers of dimensions.